### PR TITLE
Ingester refactor

### PR DESCRIPTION
**Pull request by mradamcox**
Short Description:
main_refactor.py now holds an alternative to process_instantaneous() that uses only newly refactored code. process_historical has not been migrated at all yet.

More detailed description:
Adds two new classes - agol_connection.AGOLConnection() and ingester.Ingester(). 

AGOLConnection() - loads the GIS credentials and makes a connection, and then loads some new layer configs (item id, original csv name) that are now all refactored to a json file. Once intantiated successfully, this object is set as `Ingester().gis` so it can be used for all ingestion operations.

Ingestor() - now can be used as an overarching class that runs the entire process in a step by step manner. General usage is to instantiate with dry_run true/false, and then use `Ingester().set_gis(gis)` to attach a new instance of AGOLConnection() to it. Then each of the existing steps can be made as before. Not all of the functions in `main.py` have been moved into this class yet, but the patterns have been clearly put in place and shouldn't be too difficult to pick up on, and improve.

Other notes:
- the new agol_layers.json may make more sense as a .py file like you have header_mapping, this would allow old ids to be retained with comment lines
- I added a new `--dir` argument to main_refactor.py so that you set where the sftp files are downloaded during testing.

### CODE REVIEW

File: agol_connection.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import tempfile
+import shutil
+import json
+import csv
+import os
+
+
+class AGOLConnection(object):
+
+    def __init__(self):
```
Comment (adamf) [inline agol_connection.py:12]: (not for this PR, maybe, but as a side note we should PEP-8 all this code sometime)

File: agol_connection.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        # from the published layer, you can use the FLC manager's overwrite() method successfully.
+
+        feature_item = self.gis.content.get(arcgis_item_id)
+        if "type:Table" in str(feature_item):
+            fs = feature_item.tables[0].container
+        else:
+            fs = feature_item.layers[0].container
+        return fs
+
+    def upload_to_arcgis(self, dataset_name, source_data_dir, source_data_file):
```
Comment (adamf) [inline agol_connection.py]: What do you think about renaming to `overwrite_arcgis_layer` or similar? That'd be more specific then `upload`.

File: agol_layers.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
+{
+  "feature_layer": {
```
Comment (adamf) [inline agol_layers.json]: s/feature_layer/hospital_layer/  maybe?

File: agol_layers.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (adamf) [inline agol_layers.json:1]: I like the idea of moving all of this config to a config file. In the future (post PR merge, in another branch) I think we'll want to add some more metadata tho to make it easier to understand the meaning of each entry (eg, why some have `original_file_name` and some don't.)

File: main_refactor.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python3
+import argparse
+
+import header_mapping as hm
+from operators import process_csv
+from agol_connection import AGOLConnection
+from ingester import Ingester
+
+
+def process_instantaneous(dry_run=False, datadir=None):
```
Comment (adamf) [inline main_refactor.py]: This is much cleaner :)

**Review by adamf — COMMENTED** [review]
Added some comments - tl;dr this is much cleaner, and we should keep refactoring to get the historical stuff in so we can move this to master.

File: agol_connection.py (hunk 2 above)
Comment (mradamcox) [reply agol_connection.py]: big +1 to that

File: agol_layers.json (hunk 4 above)
Comment (mradamcox) [reply agol_layers.json:1]: Yeah, I was realizing that json isn't a good format for comments, etc., so using the pattern in header_mapping.py would probably be wiser.

File: main_refactor.py (hunk 5 above)
Comment (mradamcox) [reply main_refactor.py]: Thanks, and I think sets things in a good trajectory. I had originally just put a new function in `main.py` called `process_instantenous_refactor()`, but had to abandon that due to the circular imports it required.

**Comment by mradamcox** [comment]
@adamf Added the process_historical function in main_refactor.py and this uses all methods that are now attached to Ingester. You should be able to run main_refactor instead of main now in production, however, I only tested with dry_run = True of course, and was getting those 502 errors that I mentioned.

**Comment by mradamcox** [comment]
Ok, when renaming the upload_to_arcgis method I realized there was an issue with how the gis object was being passed around, so I ended up just moving the rest of the contents of main.py into main_refactor. For all intents and purposes, this is ready to merge, as soon as some tests of overwriting the arcgis online layers have been made.

File: agol_layers.json (hunk 3 above)
Comment (mradamcox) [reply agol_layers.json]: done

File: agol_connection.py (hunk 2 above)
Comment (mradamcox) [reply agol_connection.py]: done

**Review by adamf — APPROVED** [review]

**Pull request merged**